### ISSUE TITLE
[question] msvc_compatible() is not enough sometimes?

### ISSUE THREAD

**Issue by db4** [conan-io/conan#12452]
I'm trying to use boost package from CCI with `msvc` compiler (actually I have this problem with my own packages, but boost is good for demonstrating it) . Conan fails to find the (compatible) binary:
```
conan install boost/1.75.0@ -s compiler=msvc -s compiler.version=191 -s compiler.cppstd=14 -s compiler.runtime=dynamic

Configuration:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=14
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=191
os=Windows
[options]
[build_requires]
[env]

bzip2/1.0.8: Main binary package '040d0e3492940cd304bdcec8aa9d664c2054639c' missing. Using compatible package '5be2b7a2110ec8acdbf9a1cea9de5d60747edb34'
zlib/1.2.13: Main binary package '007a990a617b175b37992961a850c5c3b07260b3' missing. Using compatible package '6cc50b139b9c3d27b3e9042d5f5372d327b3a9f7'
Installing package: boost/1.75.0
Requirements
    boost/1.75.0 from 'conancenter' - Cache
    bzip2/1.0.8 from 'conancenter' - Cache
    zlib/1.2.13 from 'conancenter' - Cache
Packages
    boost/1.75.0:660f4ef44f652d2ba7f74a79487832fd00a5a488 - Missing
    bzip2/1.0.8:5be2b7a2110ec8acdbf9a1cea9de5d60747edb34 - Cache
    zlib/1.2.13:6cc50b139b9c3d27b3e9042d5f5372d327b3a9f7 - Cache

Installing (downloading, building) binaries...
ERROR: Missing binary: boost/1.75.0:660f4ef44f652d2ba7f74a79487832fd00a5a488
boost/1.75.0: WARN: Can't find a 'boost/1.75.0' package for the specified settings, options and dependencies:
- Settings: arch=x86_64, build_type=Release, compiler=msvc, compiler.cppstd=14, compiler.runtime=dynamic, compiler.runtime_type=Release, compiler.version=191, os=Windows
- Options: asio_no_deprecated=False, buildid=None, bzip2=True, debug_level=0, diagnostic_definitions=False, error_code_header_only=False, extra_b2_flags=None, filesystem_no_deprecated=False, filesystem_version=None, header_only=False, i18n_backend=deprecated, i18n_backend_iconv=off, i18n_backend_icu=False, layout=system, lzma=False, magic_autolink=False, multithreading=True, namespace=boost, namespace_alias=False, numa=True, pch=True, python_executable=None, python_version=None, segmented_stacks=False, shared=False, system_no_deprecated=False, system_use_utf8=False, visibility=hidden, without_atomic=False, without_chrono=False, without_container=False, without_context=False, without_contract=False, without_coroutine=False, without_date_time=False, without_exception=False, without_fiber=False, without_filesystem=False, without_graph=False, without_graph_parallel=True, without_iostreams=False, without_json=False, without_locale=False, without_log=False, without_math=False, without_mpi=True, without_nowide=False, without_program_options=False, without_python=True, without_random=False, without_regex=False, without_serialization=False, without_stacktrace=False, without_system=False, without_test=False, without_thread=False, without_timer=False, without_type_erasure=False, without_wave=False, zlib=True, zstd=False, bzip2:build_executable=True, bzip2:shared=False, zlib:shared=False
- Dependencies: zlib/1.2.13, bzip2/1.0.8
- Requirements: bzip2/1.Y.Z, zlib/1.Y.Z
- Package ID: 660f4ef44f652d2ba7f74a79487832fd00a5a488

ERROR: Missing prebuilt package for 'boost/1.75.0'
Use 'conan search boost/1.75.0@ --table=table.html -r=remote' and open the table.html file to see available packages
Or try to build locally from sources with '--build=boost'

More Info at 'https://docs.conan.io/en/latest/faq/troubleshooting.html#error-missing-prebuilt-package'
```
Why so? And how to debug such errors? `conan install boost/1.75.0@ -s compiler="Visual Studio" -s compiler.version=15` succeeds:
```
Configuration:
[settings]
arch=x86_64
build_type=Release
compiler=Visual Studio
compiler.runtime=MD
compiler.version=15
os=Windows
[options]
[build_requires]
[env]

Installing package: boost/1.75.0
Requirements
    boost/1.75.0 from 'conancenter' - Cache
    bzip2/1.0.8 from 'conancenter' - Cache
    zlib/1.2.13 from 'conancenter' - Cache
Packages
    boost/1.75.0:8d47612c404bfb3cc5b267708e639b6b2cda0c3a - Cache
    bzip2/1.0.8:5be2b7a2110ec8acdbf9a1cea9de5d60747edb34 - Cache
    zlib/1.2.13:6cc50b139b9c3d27b3e9042d5f5372d327b3a9f7 - Cache

Installing (downloading, building) binaries...
bzip2/1.0.8: Already installed!
bzip2/1.0.8: Appending PATH environment variable: D:\.conan\data\bzip2\1.0.8\_\_\package\5be2b7a2110ec8acdbf9a1cea9de5d60747edb34\bin
zlib/1.2.13: Already installed!
boost/1.75.0: Already installed!
boost/1.75.0: Disabled magic autolinking (smart and magic decisions)
Aggregating env generators
```

**Comment by db4** [comment]
I think I can explain now what's going on. `boost` recipe contains
```python
    def package_id(self):
        del self.info.options.i18n_backend

        if self.options.header_only:
            self.info.header_only()
            self.info.options.header_only = True
        else:
            del self.info.options.debug_level
            del self.info.options.filesystem_version
            del self.info.options.pch
            del self.info.options.python_executable  # PATH to the interpreter is not important, only version matters
            if self.options.without_python:
                del self.info.options.python_version
            else:
                self.info.options.python_version = self._python_version
```
but `msvc_compatible()` doesn't take these deletions into account and computes the wrong compatible id. It means that `msvc` compatibility is broken for any recipe that modifies options/setting info in its `package_id` method. @memsharded, can you look into this?

**Comment by memsharded** [comment]
Sorry this was not followed up back then. 
That was most likely solved in latest 1.66.0, https://docs.conan.io/en/1.66/changelog.html, closing as solved, but please re-open or create a new ticket if necessary, thanks for the feedback.
